Make PrunedLiveness ownership-friendly

atrick opened the pull request:
This establishes a sound foundation for building optimizer support for OSSA.

This includes functional changes because the APIs are now rigorous about recognizing and propagating information about reborrows and pointer escapes. These are the two areas that aren't handled correctly in our current suite of OSSA utilities.

This set of commits are included in a single PR primarily for testing purposes.
This includes new infrastructure for unit testing liveness analysis.
This also includes the fix for rdar://99874173: unreachable assert "No user in LiveWithin block"

I couldn't resolve all the liveness issues in one PR. Changing fundamental utilities means carefully auditing all the clients to make sure they handle the new behavior. So a verifier test is disabled and utilities will still bail-out in many situations where they shouldn't need to.

The next step is to add an OSSALiveness layer that handles "extended" live ranges. OSSA live ranges can be extended across copies or reborrows.

Then we'll gradually start enabling features in OwnershipOptUtils that were disabled because of bugs. And finally, we can replace most of the ad-hoc logic with a more principled approach.

The headline commit is:

Redesign PrunedLiveness APIs, introducing live ranges
First restore the basic PrunedLiveness abstraction to its original
intention. Move code outside of the basic abstraction that polutes the
abstraction and is fundamentally wrong from the perspective of the
liveness abstraction.

Most clients need to reason about live ranges, including the def
points, not just liveness based on use points. Add a PrunedLiveRange
layer of types that understand where the live range is
defined. Knowing where the live range is defined (the kill set) helps
reliably check that arbitrary points are within the boundary. This
way, the client doesn't need to be manage this on its own. We can also
support holes in the live range for non-SSA liveness. This makes it
safe and correct for the way liveness is now being used. This layer
safety handles:

- multiple defs
- instructions that are both uses and defs
- dead values
- unreachable code
- self-loops

So it's no longer the client's responsibility to check these things!

Add SSAPrunedLiveness and MultiDefPrunedLiveness to safely handle each
situation.

atrick committed:
Redesign PrunedLiveness APIs, introducing live ranges

First restore the basic PrunedLiveness abstraction to its original
intention. Move code outside of the basic abstraction that polutes the
abstraction and is fundamentally wrong from the perspective of the
liveness abstraction.

Most clients need to reason about live ranges, including the def
points, not just liveness based on use points. Add a PrunedLiveRange
layer of types that understand where the live range is
defined. Knowing where the live range is defined (the kill set) helps
reliably check that arbitrary points are within the boundary. This
way, the client doesn't need to be manage this on its own. We can also
support holes in the live range for non-SSA liveness. This makes it
safe and correct for the way liveness is now being used. This layer
safety handles:

- multiple defs
- instructions that are both uses and defs
- dead values
- unreachable code
- self-loops

So it's no longer the client's responsibility to check these things!

Add SSAPrunedLiveness and MultiDefPrunedLiveness to safely handle each
situation.

Split code that I can't figure out into
DiagnosticPrunedLiveness. Hopefully it will be deleted soon.

atrick committed:
Add a test case for generic specialization with borrow scope fixup.

This tests a borrow scope whose use becomes unreachable. The liveness
utilities need to work on this malformed SIL in order to fixup OSSA
on-the-fly.

atrick committed:
Update passes to use SSAPrunedLiveness or MultiDefPrunedLiveness

atrick committed:
Fix ScopedAddressValue::computeLiveness.

Return the AddressUseKind.

Fixes a bug in extendStoreBorrow where it was looking at an
uninitialized liveness result whenever a pointer escape was present.

atrick committed:
Fix findTransitiveUsesForAddress to handle incomplete scopes

This fixes ScopedAddressValue::computeLiveness in unreachable code scenarios.
For example:

   %storeBorrow = store_borrow %_ to %adr
   %loadBorrow  = load_borrow %storeBorrow
   apply %f(%loadBorrow) : $@convention(thin) (...) -> Never

Now recursively process uses of load_borrow as if they are address
uses.

Ultimately, this would be more efficiently handled by a recursive
lifetime completion utility which would fixup the load_borrow scope
before computing the store_borrow liveness.

Fixes rdar://99874173: unreachable assert "No user in LiveWithin block"

atrick committed:
ScopedAddressUtils - clarify the API used for transitive liveness.

Begin to distnguish logic that relies on complete OSSA lifetimes from
the logic that computes transitive uses.

atrick committed:
Add InnerBorrowKind and AddressUseKind to PrunedLiveness API

The API for computing simple liveness now returns a
SimpleLiveRangeSummary. Callers need to decide how to handle reborrows
and pointer escapes. If either condition exists then the resulting
liveness does not necessarily encapsulate the definition's ownership.

Fixes some number of latent bugs w.r.t. liveness clients.

atrick committed:
Add an OSSLifetimeAnalysis utility pass

Supports unit testing for OSSA lifetime utilities. These utilities are
the basis of maintaining valid OSSA form. They need to handle invalid
SIL (with incomplete OSSA lifetimes) because we rely on them to fixup
SIL after SILGen and after any transformation that may affect
ownership.

atrick committed:
Add ossa lifetime analysis tests

atrick committed:
Add OSSA liveness tests for incomplete lifetimes

These tests all involve unreachable terminators.

atrick committed:
Temporarily disable a SILVerifier unit test.

We can't verify that store borrows aren't nested until we can reliably
compute liveness.

This can be fixed in two ways, both of which we plan to do ASAP:

 (1) With complete lifetimes, this no longer needs to perform transitive
 liveness at all.

 (2) findInnerTransitiveGuaranteedUses, which ends up being called on the
 load_borrow to compute liveness, can be taught to transitively process
 InteriorPointer uses instead of returning PointerEscape. We need to make
 sure all uses of the utility need to handle this first.

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

atrick commented:
@swift-ci smoke test

atrick commented:
I'd like to merge this soon because it looks like we're getting new bug reports that it will resolve.

atrick commented on lib/SIL/Utils/ScopedAddressUtils.cpp:
```diff
@@ -86,24 +86,42 @@ bool ScopedAddressValue::visitScopeEndingUses(
+}
+
+AddressUseKind
+ScopedAddressValue::updateTransitiveLiveness(PrunedLiveness &liveness) const {
```
No, any liveness analysis may be used by a ScopedAddressValue. There's nothing specific to SSA liveness here.
```
  /// Update \p liveness for all the transitive address uses.
  ///
  /// Valid for any type of liveness, SSA or MultiDef, that may be used by a
  /// scoped address.
````

nate-chandler commented on lib/SIL/Utils/PrunedLiveness.cpp:
```diff
@@ -309,40 +227,161 @@ void PrunedLivenessBoundary::visitInsertionPoints(
+  while (searchPos) {
+    if (isLive) {
+      nextDef = asImpl().findPreviousDef(searchPos, nextDef);
+      if (!nextDef)
```
When there's only a single def (SSAPrunedLiveness), does continuing the search above that single def result in visiting instructions which can't be on the boundary?

atrick replied:
No, the main goal of this design was to avoid visiting any extra instructions for the common SSAPrunedLiveness case. That's why `findPreviousDef` is delegated to the subclass. Does `SSAPrunedLiveness::findPreviousDef` look right?

nate-chandler replied:
In an example like
```
entry:
    ...
    instructions
    ...
    %def = apply
    ...
    apply(%def) // last use
    ...
    %retval = tuple ()
    return %retval

```

It looks like we'll do the following:
1. set `isLive = false`, `isDefBlock = true`, `nextDef = nullptr`
2. advance searchPos from the end of the block toward the front.  for the first instructions, `isLive` will be false, and since `isDefBlock` is true, we'll check whether instructions are the def.  none of them will be.
3. when searchPos becomes `apply(%def)`; `isInterestingUser` will return something other than NonUser so we'll set `isLive = true`
4. enter the `isLive` condition.  call `findPreviousDef(theLastUse, nullptr)`.  we'll get back `%def = apply`.
5. set `searchPos` to `nextDef`.  set `isLive` to false.
6. keep iterating over the instructions above `%def = apply`.

At least it looks like that's what I'm seeing in the debugger for the following:

```
class K {}

sil [ossa] @getK : $@convention(thin) () -> @owned K

sil [ossa] @borrowK : $@convention(thin) (@guaranteed K) -> ()

sil [ossa] @takeK : $@convention(thin) (@owned K) -> ()

sil [ossa] @f : $() -> () {
  %getK = function_ref @getK : $@convention(thin) () -> @owned K
  %borrowK = function_ref @borrowK : $@convention(thin) (@guaranteed K) -> ()
  %takeK = function_ref @takeK : $@convention(thin) (@owned K) -> ()
  %k1 = apply %getK() : $@convention(thin) () -> @owned K
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  %k2 = apply %getK() : $@convention(thin) () -> @owned K
  debug_value [trace] %k2 : $K
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %takeK(%k2) : $@convention(thin) (@owned K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  apply %takeK(%k1) : $@convention(thin) (@owned K) -> ()
  %retval = tuple ()
  return %retval : $()
}
```

atrick replied:
@nate-chandler right! I'm glad you spotted that because I changed my mind after writing it.

/cc @eeckstein 
I intentionally continued looking for a liveness boundary in the def block handle to handle self loops:

```
bb1(%phi : @owned $K):
  apply %borrowK(%phi)
  destroy_value %phi : $K
  %k = apply %getK() : ()-> @owned $K
  br bb1(%k : $K)
```

I was concerned that, when `bb1` becomes unreachable, the `%phi` will be immediately removed by a simple peephole:

```
bb1:
  borrowK(%k) : $K
  %k = apply %getK() : ()-> @owned $K
  br bb1
```

But, I was never able to come up with a way to test this. Our SIL verifier claims that it does not meet SSA dominance requirements. In theory, it does meet SSA requirements because unreachable blocks are in fact dominated by everything in the function. In fact, introducing a meaningless branch makes it valid SIL again:

```
bb1:
  apply %borrowK(%k1) : $@convention(thin) (@guaranteed K) -> ()
  br bb2

bb2:
  %k1 = apply %getK() : $@convention(thin) () -> @owned K
  br bb1
```

But we care about the case where the def and use are in the same block. Since we can't test this, I came up (or rediscovered) a new rule:

_Redundant phis can only be eliminated after first eliminating all unreachable code_

I'll add a comment about that optimizer rule and make this liveness check more efficient.

atrick replied:
@nate-chandler @meg-gupta 
Now that I realize we don't need to guard against self-loops with SSA liveness, a big pile of complexity that I added to deal with that corner case goes away:
https://github.com/apple/swift/pull/61371/commits/281b0ce32ab6e06831bcb053f0fc62c78ef670e2

nate-chandler replied:
Hooray--thank you!

atrick committed:
Fix a comment merge error

atrick committed:
Add a clarifying comment to updateTransitiveLiveness

atrick committed:
Rename areUsesWithinTransitiveScope to ...ExtendedScope

Start using consistent terminolfy in ownership utils.

A transitive use set follows transitive uses within an ownership
lifetime. It does not rely on complete inner scopes. An extended use
set is not necessarilly transitive but does look across
lifetime-ending uses: copies of owned values and/or reborrows of
guaranteed values. Whether lifetime extension refers to copies or
reborrow is context dependent.

atrick committed:
Make PrunedLiveness::computeSimple one-level transitive

Computing simple liveness is distinct from computing transitive
liveness. But for safety and consistency, always handle the first
level of liveness transitively. This way, computeSimple can be used on
guaranteed values that need OSSA lifetime fixup.

Simple liveness just means that *inner* borrow and address scopes are
assumed to be complete.

This utility is still only used conservatively because OSSA lifetime
completion is not yet enabled. But, in the future, computeSimple can
be used in the common case.

atrick committed:
Add a comment on liveness terminology

atrick committed:
Add a transitive liveness test case

atrick commented on lib/SIL/Utils/PrunedLiveness.cpp:
```diff
@@ -309,40 +243,169 @@ void PrunedLivenessBoundary::visitInsertionPoints(
+      break;
+    case OperandOwnership::InteriorPointer:
+      checkAndUpdateInteriorPointer(use);
+      break;
```
> don't we have to explore interior pointer uses transitively via findTransitiveUsesForAddress ?

I intentionally did not make `computeSimple` a transitive analysis. It was not transitive before this PR and is currently only used conservatively. Note that recognizing scoped uses is always needed even if you are not computing transitive uses.

But, in the interest of general safety, I agree that the first level only should actually be transitive. That way `computeSimple` makes the most sense for guaranteed values when using the liveness result to reason about ownership.

I added a test case and comments on the terminology. There are 5 new commits in this PR to handle this change.

meg-gupta replied:
Thanks!

meg-gupta commented on include/swift/SIL/OwnershipUtils.h:
```diff
@@ -9,6 +9,36 @@
+/// - Reborrow-extension is used to check borrow scopes relative to its inner
+///   uses and outer lifetime
+/// - Per-definition dominance does not hold
+///
```
Thanks for writing this down

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

meg-gupta commented:
LGTM

nate-chandler reviewed (approved):
Other than the one question I'm still unclear on, LGTM too

atrick committed:
Add a comment regarding an optimizer rule for self-loops.

atrick committed:
Remove complexity in findBoundariesInBlock

We can remove a ton of complexity by assuming that SSA uses never
occur before a def in the same block. This will hold true as long as
useless phis are only removed after all unreachable code is first
removed. Then we don't have the self-loop problem.

The SILVerifier already checks this invariant and I added an in-depth
comment in SimplifyCFG.

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

atrick commented:
@swift-ci smoke test and merge
